<html>
<head>
<meta charset="utf-8"/>
<meta content="width=device-width, initial-scale=1.0, maximum-scale=1.0" name="viewport"/>
<title>Jake's CS Notes - OOP</title>
<link href="https://fonts.googleapis.com/css?family=Inconsolata" rel="stylesheet"/>
<link href="../css/testStyle.css" rel="stylesheet"/>
<link href="../css/notePageStyle.css" rel="stylesheet"/>
<link href="../css/cs1331Theme.css" id="class-theme-styles" rel="stylesheet"/>
<link crossorigin="anonymous" href="https://use.fontawesome.com/releases/v5.0.13/css/all.css" integrity="sha384-DNOHZ68U8hZfKXOrtjWvjxusGo9WQnrNx2sqG0tfsghAvtVlRW3tvkXWZh58N9jp" rel="stylesheet"/>
</head>
<body>
<script defer="" src="../js/wrapText.js"></script>
<script defer="" src="../js/pageTransitions.js"></script>
<nav class="nav-top">
<ul>
<li class="link-with-slash"><a href="../index.html"><i class="fas fa-home"></i></a></li>
<li><a href="#" id="class-title-link">OOP</a></li>
</ul>
<ul class="note-links-slider"><li><a class="is-note-link" href="0_classNotes_9_14_16.html">0. Classes</a></li><li><a class="is-note-link" href="1_moreClassNotes_9_16_16.html">1. Classes (Cont.)</a></li><li><a class="is-note-link" href="2_inheritanceNotes_9_26_16.html">2. Inheritance</a></li><li><a class="is-note-link" href="3_moreInheritanceNotes_9_28_16.html">3. Inheritance (cont.)</a></li><li><a class="is-note-link" href="4_polymorphismNotes_9_30_16.html">4. Polymorphism</a></li><li><a class="is-note-link" href="5_morePolymorphismNotes_10_3_16.html">5. Polymorphism (cont.)</a></li><li><a class="is-note-link" href="6_objectSuperclassNotes_10_5_16.html">6. The Object Superclass</a></li><li><a class="is-note-link" href="7_moreObjectSuperclassNotes_10_7_16.html">7. Object Superclass (cont.)</a></li><li><a class="is-note-link" href="8_exceptionNotes_10_19_16.html">8. Exceptions</a></li><li class="active-note-page"><a class="is-note-link" href="9_moreExceptionNotes_10_21_16.html">9. Exceptions (cont.)</a></li><li><a class="is-note-link" href="10_collectionsNotes_10_24_16.html">10. Java Collections</a></li><li><a class="is-note-link" href="11_moreCollectionNotes_10_26_16.html">11. Collections (cont.)</a></li><li><a class="is-note-link" href="12_collectionAlgorithms_10_31_16.html">12. Collection Algorithms</a></li><li><a class="is-note-link" href="13_iteratorNotes_11_2_16.html">13. Iterators</a></li><li><a class="is-note-link" href="14_lambdaNotes_11_4_16.html">14. Lambda Expressions</a></li><li><a class="is-note-link" href="15_hashedNotes_11_7_16.html">15. Hashed Collections</a></li><li><a class="is-note-link" href="16_javafxNotes_11_9_16.html">16. JavaFX Notes</a></li><li><a class="is-note-link" href="17_JavaFXNotes_11_11_16.html">17. JavaFX (cont.)</a></li><li><a class="is-note-link" href="18_test3Review_11_14_16.html">18. Test 3 Review</a></li><li><a class="is-note-link" href="19_recursionNotes_11_28_16.html">19. Recursion</a></li><li><a class="is-note-link" href="20_linkedListNotes_11_30_16.html">20. Linked Lists</a></li><li><a class="is-note-link" href="21_stackQueueNotes_12_2_16.html">21. Stacks &amp; Queues</a></li><li><a class="is-note-link" href="22_closingNotes_12_5_16.html">22. Closing Time</a></li></ul>
</nav>
<main>
<a class="side-link is-note-link" href="8_exceptionNotes_10_19_16.html"></a>
<article>
<!-- Actual note text goes into 'pre' -->
<span id="text-width-ruler"></span>
<pre class="main-note-text">//************************************************//
//***Exceptions (cont.) - October 21st, 2016*****//
//**********************************************//

- (cont.) So, to reiterate, runtime exceptions are "unchecked exceptions" that the compiler can't catch, whereas "checked exceptions" are checked before while compiling
    -Exceptions themselves do NOT actually happen at compile time; think of it as return types, where it just checks to see if the return types match up
    -e.g. if we write a that method that throws an exception, even though nothing went wrong, the compiler will freak out (?)

-Main way to handle exceptions is the "try {...} catch (Exception x) {...}" block
    -CAN HAVE MULTIPLE CATCH BLOCKS to catch multiple types of exceptions

-adding "throws [ExceptionTypeX]" to end of method header lets it throw exceptions
    -Once thrown, checks to see if the local method can handle it; otherwise, it throws it up to whatever called the method, and the process continues until the exception is either handled, or there's nothing else to go up to, at which point the program crashes.
    -WILL NOT COMPILE if the program does not either handle or throw the exception (if it's a checked exception)

-Good method to know: printStackTrace()

-If the code is handled, then the program won't crash without being explicitly told to; it will continue to run (might exit the current method? Goes to catch statement and resumes control from the end of that, I think?)

public class Wee {
    static void bar() throws Throwable {
        throw new Throwable("wee!");
    }

    static void foo() throws new Throwable {
        bar();
        System.out.println("Foo!");
    }

    public static void main(String[] args) {
        try {
            foo();
        } catch (Throwable t) {
            System.out.println(t.getMessage);
        }
        System.out.println("Still running!");
    }
}

-This code will print out "wee! Still running!", since foo() immediately calls bar(), which throws an exception; since it can't handle it, it throws it up to foo (where it was called from), and then foo throws it up to main, which then moves control to the 'catch' statement; the catch statement prints the Throwable's message (which happens to be "wee!"), and the program then picks up from the end of the catch statement and print 'Still running!';
    -If we have an exception IN the main method (not just a method called in main), then there's nowhere to throw it, so the program will crash

-We don't have to write "catch" statements for unchecked/runtime exceptions in order for the program to compile; HOWEVER, we CAN write "try...catch" blocks for them in THE EXACT SAME WAY!!!
    -EXTREMELY useful for checking user input; theoretically, everything else should be already written fine by you, but the compiler won't stop you from checking for exceptions elsewhere

-Important things when writing more than 1 catch statement:
    -The exception type in a catch clause matches up w/ the subclass
    -The first catch clause to match the thrown exception should be the only one to execute
-So, basically, should go from most-to-least specific as you go down

-You can write your own exceptions my extending the "Exception" class
    -Don't do this unless you really need to, though; the Java library is HUGE, and 99.9% of the errors you would want to check for have already been written by some experts who've done the heavy lifting for you

-Don't use too many exceptions; if you have to use them all over the place, it means you probably are writing sloppy code
-You should only REALLY use exceptions when you're dealing with data that you can't control, like user input

-It's helpful to go through and see what compiles and what causes errors; that's really the best way to understand how these things work
</pre>
</article>
<a class="side-link is-note-link" href="10_collectionsNotes_10_24_16.html"></a>
</main>
</body>
</html>